src/compiler/scala/tools/nsc/PhaseAssembly.scala

```diff
@@ -226,50 +232,92 @@ object DependencyGraph {
+    def addConstraint(from: String, to: String, weight: Weight): Unit = constraints.addOne((from, to, weight))
+    def checkConstraint(p: SubComponent)(name: String, constraint: String, warn: Boolean): Boolean = isPhaseName(name) || {
+      if (warn) {
+        val knownNames = phaseNames ++ compilerPhases
+        val help = if (knownNames.contains(name)) "" else
+          knownNames.filter(util.EditDistance.levenshtein(name, _) < 3).toList match {
+            case Nil => ""
+            case close => s" - did you mean ${util.StringUtil.oxford(close, "or")}?"
+          }
+        warnings.addOne(s"No phase `$name` for ${p.phaseName}.$constraint$help")
       }
-    for (p <- phases) {
-      require(p.phaseName.nonEmpty, "Phase name must be non-empty.")
-      def checkConstraint(name: String, constraint: String): Boolean =
-        graph.components.contains(name).tap(ok => if (!ok && warn) graph.warning(s"No phase `$name` for ${p.phaseName}.$constraint${phaseTypo(name)}"))
-      for (after <- p.runsRightAfter if after.nonEmpty && checkConstraint(after, "runsRightAfter"))
-        graph.addEdge(after, p.phaseName, FollowsNow)
-      for (after <- p.runsAfter if after.nonEmpty && !p.runsRightAfter.contains(after) && checkConstraint(after, "runsAfter"))
-        graph.addEdge(after, p.phaseName, Follows)
-      for (before <- p.runsBefore if before.nonEmpty && checkConstraint(before, "runsBefore"))
-        graph.addEdge(p.phaseName, before, Follows)
-      // Add "runsBefore terminal" to phases without (or with invalid) runsBefore
+      false
+    }
+    for ((p, i) <- phases.zipWithIndex) {
+      for (after <- p.runsRightAfter if after.nonEmpty)
+        if (checkConstraint(p)(after, "runsRightAfter", warn=warnAll)) addConstraint(after, p.phaseName, FollowsNow)
+        else dropped += p.phaseName
+      for (after <- p.runsAfter if after.nonEmpty && !p.runsRightAfter.contains(after))
+        if (checkConstraint(p)(after, "runsAfter", warn=warnAll)) addConstraint(after, p.phaseName, Follows)
+        else dropped += p.phaseName
+      for (before <- p.runsBefore if before.nonEmpty)
+        if (checkConstraint(p)(before, "runsBefore", warn=warnAll)) addConstraint(p.phaseName, before, Follows)
+      // terminal follows every phase; parser is not before every phase because traverse uses edgeTo to find "anchors"
       if (p != end || p == end && p == start)
-        if (!p.runsBefore.exists(graph.components.contains))
-          graph.addEdge(p.phaseName, end.phaseName, Follows)
+        addConstraint(p.phaseName, end.phaseName, Follows)
+    }
+    if (warnAll)
+      dropped.foreach(p => warnings.addOne(s"Component ${p} dropped due to bad constraint"))
+    val purgedConstraints = constraints.filterInPlace {
+      case (from, to, w) => !dropped.contains(from) && !dropped.contains(to)
+    }.toList
+    new DependencyGraph(start.phaseName, phases.iterator.map(p => p.phaseName -> p).toMap).tap { graph =>
```

Comment:
what's the advantage of the new approach of collecting constraints before creating the `DependencyGraph`?